Fixup for PR10896: Avoid old compiler bug

Geod24 opened the pull request:
```
The static if would trigger a forward reference bug which would corrupt
the vtable of other aggregates.
Using a template constraint changes the stage at which semantic is performed,
avoiding the bug altogether.
Note that the bug was fixed in v2.085 but compiling with older compilers is sometimes needed.
```

Discussion: https://github.com/dlang/dmd/pull/10896#discussion_r392059295
CC @ibuclaw @kinke

Geod24 committed:
Fixup for PR10896: Avoid old compiler bug

The static if would trigger a forward reference bug which would corrupt
the vtable of other aggregates.
Using a template constraint changes the stage at which semantic is performed,
avoiding the bug altogether.
Note that the bug was fixed in v2.085 but compiling with older compilers is sometimes needed.

ibuclaw reviewed (changes requested):
There's a few more things that need to be addressed here.

1. The use of lambda requires access to a non-local context (issue 5710), this is not supported in either gdc, ldc, or any dmd compiled older than 2.087.0.  So all predicates need to be static functions, otherwise we'll get more compiler errors from the first commit this function is actually used.

2. The unittest for testing this function has a problem, `baguette` does not come before `Avocado` in a sorted array.

ibuclaw commented on src/dmd/root/array.d:
```diff
@@ -323,13 +323,11 @@ public:
-        {
-            return this.sort!((pe1, pe2) => pe1.opCmp(*pe2));
-        }
+        return this.sort!((pe1, pe2) => pe1.opCmp(*pe2));
```
```suggestion
        return this.sort!(function (in T* pe1, in T* pe2) => pe1.opCmp(*pe2));
```
Though I can't say why `in T*` is needed.

ibuclaw replied:
A similar change is required in the unittest too,
```
strings.sort!(function (const(char**) e1, const(char**) e2) => strcmp(*e1, *e2));
```

Geod24 commented:
> The use of lambda requires access to a non-local context

I thought DMD would infer that no context is used ? `arraySortWrapper` should error if something that isn't a `function` is passed.

> The unittest for testing this function has a problem

Right, so why is master green ? :|

ibuclaw commented:
> Right, so why is master green ? :|

Because the dmd.root.array module (and others that are in LEXER_OBJ) are not built with `-unittest`.
```
(DC) LEXER_OBJ
Run: /usr/bin/dmd -c -of../generated/linux/release/64/lexer.o -vtls -version=MARS -w -de -fPIC -m64 -J../generated/linux/release/64 -dip25 -g -color=on dmd/console.d dmd/entity.d dmd/errors.d dmd/filecache.d dmd/globals.d dmd/id.d dmd/identifier.d dmd/lexer.d dmd/tokens.d dmd/utf.d dmd/root/array.d dmd/root/bitarray.d dmd/root/ctfloat.d dmd/root/file.d dmd/root/filename.d dmd/root/hash.d dmd/root/outbuffer.d dmd/root/port.d dmd/root/region.d dmd/root/rmem.d dmd/root/rootobject.d dmd/root/stringtable.d
(DC) DMD-unittest
Run: /usr/bin/dmd -of../generated/linux/release/64/dmd-unittest -vtls -J../res -version=NoMain -unittest -main -version=MARS -w -de -fPIC -m64 -J../generated/linux/release/64 -dip25 -g -color=on dmd/access.d dmd/aggregate.d dmd/aliasthis.d dmd/apply.d dmd/argtypes.d dmd/argtypes_sysv_x64.d dmd/arrayop.d dmd/arraytypes.d dmd/ast_node.d dmd/astcodegen.d dmd/asttypename.d dmd/attrib.d dmd/blockexit.d dmd/builtin.d dmd/canthrow.d dmd/chkformat.d dmd/cli.d dmd/clone.d dmd/compiler.d dmd/complex.d dmd/cond.d dmd/constfold.d dmd/cppmangle.d dmd/cppmanglewin.d dmd/ctfeexpr.d dmd/ctorflow.d dmd/dcast.d dmd/dclass.d dmd/declaration.d dmd/delegatize.d dmd/denum.d dmd/dimport.d dmd/dinifile.d dmd/dinterpret.d dmd/dmacro.d dmd/dmangle.d dmd/dmodule.d dmd/doc.d dmd/dscope.d dmd/dstruct.d dmd/dsymbol.d dmd/dsymbolsem.d dmd/dtemplate.d dmd/dtoh.d dmd/dversion.d dmd/env.d dmd/escape.d dmd/expression.d dmd/expressionsem.d dmd/func.d dmd/hdrgen.d dmd/impcnvtab.d dmd/imphint.d dmd/init.d dmd/initsem.d dmd/inline.d dmd/inlinecost.d dmd/intrange.d dmd/json.d dmd/lambdacomp.d dmd/lib.d dmd/libelf.d dmd/libmach.d dmd/libmscoff.d dmd/libomf.d dmd/link.d dmd/mars.d dmd/mtype.d dmd/nogc.d dmd/nspace.d dmd/ob.d dmd/objc.d dmd/opover.d dmd/optimize.d dmd/parse.d dmd/parsetimevisitor.d dmd/permissivevisitor.d dmd/printast.d dmd/safe.d dmd/sapply.d dmd/scanelf.d dmd/scanmach.d dmd/scanmscoff.d dmd/scanomf.d dmd/semantic2.d dmd/semantic3.d dmd/sideeffect.d dmd/statement.d dmd/statement_rewrite_walker.d dmd/statementsem.d dmd/staticassert.d dmd/staticcond.d dmd/target.d dmd/templateparamsem.d dmd/traits.d dmd/transitivevisitor.d dmd/typesem.d dmd/typinf.d dmd/utils.d dmd/visitor.d dmd/vsoptions.d dmd/foreachvar.d dmd/irstate.d dmd/toctype.d dmd/glue.d dmd/gluelayer.d dmd/todt.d dmd/tocsym.d dmd/toir.d dmd/dmsc.d dmd/tocvdebug.d dmd/s2ir.d dmd/toobj.d dmd/e2ir.d dmd/eh.d dmd/iasm.d dmd/iasmdmd.d dmd/iasmgcc.d dmd/objc_glue.d dmd/backend/cc.d dmd/backend/cdef.d dmd/backend/cgcv.d dmd/backend/code.d dmd/backend/cv4.d dmd/backend/dt.d dmd/backend/el.d dmd/backend/global.d dmd/backend/obj.d dmd/backend/oper.d dmd/backend/outbuf.d dmd/backend/rtlsym.d dmd/backend/code_x86.d dmd/backend/iasm.d dmd/backend/codebuilder.d dmd/backend/ty.d dmd/backend/type.d dmd/backend/exh.d dmd/backend/mach.d dmd/backend/mscoff.d dmd/backend/dwarf.d dmd/backend/dwarf2.d dmd/backend/xmm.d dmd/backend/dlist.d dmd/backend/melf.d dmd/root/aav.d dmd/root/longdouble.d dmd/root/man.d dmd/root/response.d dmd/root/speller.d dmd/root/string.d dmd/root/strtold.d ../generated/linux/release/64/lexer.o ../generated/linux/release/64/backend.o
```

ibuclaw commented:
> I thought DMD would infer that no context is used ? arraySortWrapper should error if something that isn't a function is passed.

I'm building gdc-10-ddmd with gdc-9, and I get errors with `-funittest`.
```
gdc -fno-PIE -c  -g -O2 -funittest -fversion=IN_GCC  -fPIC   -Wall -Wdeprecated -fno-common -o d/arrayop.o -MT d/arrayop.o -MMD -MP -MF d/.deps/arrayop.TPo -I../../gcc/d -J../../gcc/d/dmd -J../../gcc/d/dmd/ddoc ../../gcc/d/dmd/arrayop.d
../../gcc/d/dmd/root/array.d:540:12: error: template instance sort!((e1, e2) => strcmp(*e1, *e2)) cannot use local '__lambda1' as parameter to non-global template sort(alias pred)()
  540 |     strings.sort!((e1, e2) => strcmp(*e1, *e2));
      |            ^
../../gcc/d/dmd/root/array.d:332:24: error: template instance sort!((pe1, pe2) => pe1.opCmp(*pe2)) cannot use local '__lambda1' as parameter to non-global template sort(alias pred)()
  332 |             return this.sort!((pe1, pe2) => pe1.opCmp(*pe2));
      |                        ^
../../gcc/d/dmd/root/array.d:564:14: error: template instance dmd.root.array.Array!(MyStruct).Array.sort!() error instantiating
  564 |     arr1.sort();
      |              ^
```

ibuclaw commented:
@Geod24 I guess you're trying to fix the build tool then so that unittests are compiled? :-)

Geod24 commented:
Was a bit busy this week, but yes, I will do that soon :)

Geod24 committed:
build.d: Pass -unittest to the lexer and backend too

Otherwise unittests in those modules will never be run.
Also correct the broken unittests.

Geod24 commented:
@ibuclaw : Fixed.
@MoonlightSentinel : Can you check the second commit ?

ibuclaw commented on src/build.d:
```diff
@@ -445,7 +448,7 @@ alias runCxxUnittest = makeRule!((runCxxBuilder, runCxxRule) {
         .description("Build the C++ unittests")
         .msg("(DMD) CXX-UNITTEST")
-        .deps([lexer, backend, cxxFrontend])
+        .deps([lexer(null), backend(null), cxxFrontend])
```
Were does `lexer(null)` get written to? Does it overwrite the previously built lexer archive?

MoonlightSentinel replied:
No, these targets use `memoize` to create and store unique instances.

Geod24 replied:
I think the comment is about the actual binary, and valid. For example `dmdExe`'s first argument is there for this very reason.

MoonlightSentinel replied:
I see. Agreed on the binary issue, see https://github.com/dlang/dmd/pull/10930#pullrequestreview-381013711

MoonlightSentinel reviewed (changes requested):
This changes creates a race condition between `unittest` and other targets depending on `lexer` and `backend`, (e.g. `./build.d unittest cxx-unittest`) because all dependencies write the same binary.

There needs to be a unique filename for every instance, e.g. `lexer-unittest.$(OBJ)`

Geod24 commented:
This should finally be good to go.

MoonlightSentinel commented on src/build.d:
```diff
@@ -393,10 +398,12 @@ alias dmdExe = makeRuleWithArgs!((MethodInitializer!BuildRule builder, BuildRule
         .target(env["DMD_PATH"] ~ targetSuffix)
         .msg("(DC) DMD" ~ targetSuffix)
-        .deps([versionFile, sysconfDirFile, lexer, backend])
+        .deps([versionFile, sysconfDirFile, lexer(targetSuffix, depFlags), backend(targetSuffix, depFlags)])
```
Maybe store the concrete `lexer`/`backend` instance in a variable?

MoonlightSentinel commented on src/build.d:
```diff
@@ -260,9 +260,11 @@ alias autoTesterBuild = makeRule!((builder, rule) {
+    .target(env["G"].buildPath("lexer" ~ suffix).objName)
     .sources(sources.lexer)
     .deps([versionFile, sysconfDirFile])
     .msg("(DC) LEXER_OBJ")
```
Different `lexer` targets should display different messages. Ditto for `backend`,
```suggestion
    .msg("(DC) LEXER" ~ suffix)
```

Geod24 commented:
Addressed review

MoonlightSentinel reviewed (approved):
The changes to `build.d` seem fine.

thewilsonator commented:
@Geod24 Might want to give it a force push or rebase, doc tester has failed. Not sure why.

Geod24 commented:
Just a timeout. Rebased.
